core/src/main/java/com/amplifyframework/auth/options/AuthSignUpOptions.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.auth.options;
+
+import androidx.annotation.NonNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public final class AuthSignUpOptions {
+    private final Map<String, String> userAttributes;
+    private final Map<String, String> validationData;
+
+    private AuthSignUpOptions(
+            Map<String, String> userAttributes,
+            Map<String, String> validationData
+    ) {
+        this.userAttributes = userAttributes;
+        this.validationData = validationData;
+    }
+
+    @NonNull
+    public Map<String, String> getUserAttributes() {
+        return userAttributes;
+    }
+
+    @NonNull
+    public Map<String, String> getValidationData() {
+        return validationData;
+    }
+
+    @NonNull
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+        private Map<String, String> userAttributes;
+        private Map<String, String> validationData;
+
+        public Builder() {
+            this.userAttributes = new HashMap<>();
+            this.validationData = new HashMap<>();
+        }
+
+        @NonNull
+        public Builder userAttributes(@NonNull Map<String, String> userAttributes) {
```

Comment:
This is like the Analytics API. When you provide a map, the user can do stuff like:
```
Map<String, String> userAttributes = new HashMap<>();
map.put("emailZZ", "jameson@domain.tld");
```

And they won't realize they mispelled anything until their app is out in the field and it fails at runtime.

So, data modeling can help.

`Map<String, Object>` is basically a `class`, where you name the field members in keys, and populate their values in values.

`Map<String, String>` is more or less the same thing, but with less intricate data types.

For example,

```
Map<String, String> userAttributes = new HashMap<>();
enum AttributeKeys {
    EMAIL("email"),
    NAME("name"); // .... etc.. ..
}
interface Whatever {
    // By the way, please make sure that, by convention, you use AttributeKeys!!!
    void useAttributes(Map<String, String> userAttributes);
}
```
Well, okay. But we could have done:
```
final class UserAttributes {
    private Email email;
    private Name name;
   
    UserAttributes(Email email, Name name) {
       this.email = email;
       this.name = name;
    }

    static final class Email {
         private final String validEmail;
         private Email(String validEmail) {
            this.validEmail = validEmail; 
         }
         public static Email parse(String anyString) {
             // Emails have a defined format with TLD at the end, an @ in the middle, etc....
             // Does this string?
             return new Email(/* validated */ anyString);
         }
         public String getUserName() {
             // reutrn the part before at symbol
         }
         public String getDomainName() {
             // return portion after at...
         }
    }

    static final class Name {
        private final String validatedName;
        private Name(String validatedName) {
            this.validatedName = validatedName);
        }
   
        public static Name parse(@Nullable String fullName) {
            // Validate the format of the string is actually a name....
            // For sake of argument, let's require a first and last name.
            return new Name(/*valid! */ string);
        }

        public String getFirstName() {
        
    }
}
```

Okay so anyway, my point is, that data modeling created a semantic experience for a user, that is a lot more fool proof. The user would make this stuff like:

````
// This can be a step builder, to catch all the required fields, and get that stuff at runtime.
UserAttributes.builder()
    .email(Email.parse("tony@koolbiz.uk"))
    .name(Name.parse("Tony Daniels"))
    .build();
````

Compare with:
```
Map<String, String> attrs = new HashMap<>();
attrs.put("Eamil", "whoops");
attrs.put("Eamil", "whoopsOverwrite");
```

Alot more of the stuff is left to reading the docs carefully and then validating at runtime.

Reply:
So the thing with this is that it's intentionally left open because the customer needs to be able to pass a map of any key/value pairs. There's no limit on what they could be.